### PR TITLE
Async Norbert Handler Draft

### PR DESCRIPTION
**Pull request by raver89**

### CODE REVIEW

File: java-network/src/main/java/com/linkedin/norbert/javacompat/network/CallbackRequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.linkedin.norbert.javacompat.network;
+
+import com.linkedin.norbert.network.server.CallbackContext;
+import java.util.function.BiConsumer;
```
Comment (calcu16) [inline java-network/src/main/java/com/linkedin/norbert/javacompat/network/CallbackRequestHandler.java]: Are we actually using this, I don't see it

File: network/src/main/scala/com/linkedin/norbert/network/common/LocalMessageExecution.scala [hunk 2]
```diff
@@ -27,7 +27,7 @@ trait LocalMessageExecution extends BaseNetworkClient {
 
   override protected def doSendRequest[RequestMsg, ResponseMsg](requestCtx: Request[RequestMsg, ResponseMsg])
   (implicit is: InputSerializer[RequestMsg, ResponseMsg], os: OutputSerializer[RequestMsg, ResponseMsg]): Unit = {
-    if (requestCtx.node == myNode) messageExecutor.executeMessage(requestCtx.message, requestCtx.callback)
+    if (requestCtx.node == myNode) messageExecutor.execMSG(requestCtx.message, requestCtx.name, requestCtx.callback)
```
Comment (calcu16) [inline network/src/main/scala/com/linkedin/norbert/network/common/LocalMessageExecution.scala]: Was executeMessage really too long?

File: network/src/main/scala/com/linkedin/norbert/network/server/MessageHandlerRegistryComponent.scala [hunk 3]
```diff
@@ -21,47 +21,73 @@ trait MessageHandlerRegistryComponent {
...
                                               (implicit is: InputSerializer[RequestMsg, ResponseMsg], os: OutputSerializer[RequestMsg, ResponseMsg]) {
     if (handler == null) throw new NullPointerException
 
-    handlerMap += (is.requestName -> MessageHandlerEntry(is, os, handler))
+    syncHandlerMap += (is.requestName -> SyncHandlerEntry(is, os, handler))
   }
 
+  def registerAsyncHandler[RequestMsg, ResponseMsg](handler: (RequestMsg, CallbackContext[ResponseMsg]) => Unit)
+                                                   (implicit is: InputSerializer[RequestMsg, ResponseMsg], os: OutputSerializer[RequestMsg, ResponseMsg]) {
+    if (handler == null) throw new NullPointerException
```
Comment (calcu16) [inline network/src/main/scala/com/linkedin/norbert/network/server/MessageHandlerRegistryComponent.scala]: (style) parens

File: network/src/test/scala/com/linkedin/norbert/network/server/MessageExecutorSpec.scala [hunk 4]
```diff
@@ -65,128 +65,182 @@ class MessageExecutorSpec extends SpecificationWithJUnit with Mockito with WaitF
     "satisfy general requirements in" in { generalExecutorTests }
 
     "satisfy filter-specific requirements in" in { filterSpecificTests }
-
-//    "execute the responseHandler with Left(InvalidMessageException) if the response message is of the wrong type" in {
```
Comment (calcu16) [inline network/src/test/scala/com/linkedin/norbert/network/server/MessageExecutorSpec.scala]: I'd rather this just be deleted if it doesn't make sense any more

File: network/src/main/scala/com/linkedin/norbert/network/server/MessageHandlerRegistryComponent.scala [hunk 5]
```diff
@@ -21,47 +21,73 @@ trait MessageHandlerRegistryComponent {
   val messageHandlerRegistry: MessageHandlerRegistry
 }
 
-private case class MessageHandlerEntry[RequestMsg, ResponseMsg]
+private case class SyncHandlerEntry[RequestMsg, ResponseMsg]
```
Comment (sungjuc) [inline network/src/main/scala/com/linkedin/norbert/network/server/MessageHandlerRegistryComponent.scala]: Maybe we can introduce generic HandlerEntry with case classes that can keep code more cleaner. Inside of the MessageHanlderRegistryComponent, try to handle selecting sync and async cases as much as possible to keep client side. Thus, we can also remove to check whether a given message is sync or async types.

File: java-network/src/main/java/com/linkedin/norbert/javacompat/network/CallbackRequestHandler.java (hunk 1 above)
Comment (raver89) [inline java-network/src/main/java/com/linkedin/norbert/javacompat/network/CallbackRequestHandler.java]: fixed

File: network/src/main/scala/com/linkedin/norbert/network/common/LocalMessageExecution.scala (hunk 2 above)
Comment (raver89) [inline network/src/main/scala/com/linkedin/norbert/network/common/LocalMessageExecution.scala]: Changed it back

**Pull request merged**